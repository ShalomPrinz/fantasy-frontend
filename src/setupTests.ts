// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import axios from "axios";
import { setupServer } from "msw/node";

import { handlers } from "../test";
import * as TeamContext from "./contexts/TeamContext";
import * as UserContext from "./features/authentication/contexts/UserContext";
import { LeagueInfo, Message, Player, Team, User } from "./types";

jest.setTimeout(10000);

axios.defaults.baseURL = "";

const server = setupServer(...handlers);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

interface MockedUserContext {
  logout?: () => Promise<void>;
  state?: UserContext.UserState;
  user?: User;
}

interface MockedUser {
  inbox?: Message[];
  leagues?: LeagueInfo[];
  name?: string;
  players?: Player[];
}

export const getTestUser = (user?: MockedUser): User => ({
  id: "id",
  inbox: user?.inbox || [],
  leagues: user?.leagues || [],
  name: user?.name || "",
  team: new Team(user?.players || []),
});

export const mockUser = (context: MockedUserContext) => {
  const userContext = jest.spyOn(UserContext, "useUser");

  const { logout, state, user } = context;
  const mockedLogout = typeof logout !== "undefined" ? logout : async () => {};
  const mockedState =
    typeof state !== "undefined" ? state : UserContext.UserState.LOGGED_USER;
  const mockedUser = typeof user !== "undefined" ? user : getTestUser();

  userContext.mockImplementation(() => ({
    loading: false,
    login: async () => {},
    logout: mockedLogout,
    register: async () => {},
    state: mockedState,
    user: mockedUser,
    invalidate: async () => {},
  }));

  return userContext.mockRestore;
};

export const mockTeam = (team: Team) => {
  const teamContext = jest.spyOn(TeamContext, "useTeamState");
  teamContext.mockImplementation(() => team);
};

export * from "../test";
