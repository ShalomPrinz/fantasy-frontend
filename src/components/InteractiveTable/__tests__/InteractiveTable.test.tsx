import { clickElement, render, screen } from "setupTests";

import { InteractiveTable } from "../";

const columns = [
  {
    id: 0,
    label: "First Column",
    path: "name",
  },
];

const firstRowData = {
  id: 0,
  name: "Test Name",
};

const data = [firstRowData];

describe("InteractiveTable", () => {
  it("should render InteractiveTable component by path", () => {
    const { asFragment } = render(
      <InteractiveTable columns={columns} data={data} onRowClick={jest.fn()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render InteractiveTable component by content", () => {
    const columns = [
      {
        id: 0,
        label: "First Column",
        content: (item: any) => "name - " + item.name,
      },
    ];
    const { asFragment } = render(
      <InteractiveTable columns={columns} data={data} onRowClick={jest.fn()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render InteractiveTable with active row", () => {
    const secondRowData = {
      id: 1,
      name: "Other Name",
    };
    const twoRowsData = [...data, secondRowData];
    const isActive = (item: any) => item === secondRowData;

    const { asFragment } = render(
      <InteractiveTable
        columns={columns}
        data={twoRowsData}
        isActive={isActive}
        onRowClick={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe("When table row is clicked", () => {
    it("should call given onRowClick", async () => {
      const onRowClick = jest.fn();
      const { user } = render(
        <InteractiveTable
          columns={columns}
          data={data}
          onRowClick={onRowClick}
        />
      );

      const tableRow = screen.getAllByRole("row")[1];
      await clickElement(user, tableRow);
      expect(onRowClick).toBeCalledWith(firstRowData);
    });
  });
});
