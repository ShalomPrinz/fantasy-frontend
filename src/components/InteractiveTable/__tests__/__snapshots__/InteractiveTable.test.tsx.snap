// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractiveTable should render InteractiveTable component by content 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white table-hover fs-5 text-center"
  >
    <thead>
      <tr>
        <th
          class="p-3"
        >
          First Column
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="clickable"
      >
        <td
          class="align-middle"
        >
          name - Test Name
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`InteractiveTable should render InteractiveTable component by path 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white table-hover fs-5 text-center"
  >
    <thead>
      <tr>
        <th
          class="p-3"
        >
          First Column
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="clickable"
      >
        <td
          class="align-middle"
        >
          Test Name
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`InteractiveTable should render InteractiveTable with active row 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white table-hover fs-5 text-center"
  >
    <thead>
      <tr>
        <th
          class="p-3"
        >
          First Column
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="clickable"
      >
        <td
          class="align-middle"
        >
          Test Name
        </td>
      </tr>
      <tr
        class="clickable bg-info"
      >
        <td
          class="align-middle"
        >
          Other Name
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
